ratis-server/src/main/java/org/apache/ratis/server/impl/StateMachineUpdater.java

```diff
@@ -190,11 +189,10 @@ public void run() {
           reload();
         }
 
-        final MemoizedSupplier<List<CompletableFuture<Message>>> futures = applyLog();
-        checkAndTakeSnapshot(futures);
+        applyLogFutures = applyLog(applyLogFutures);
+        checkAndTakeSnapshot(applyLogFutures);
 
         if (shouldStop()) {
-          checkAndTakeSnapshot(futures);
```

Comment:
We need `applyLogFutures.get()`.  Otherwise, it could stop without waiting the future.

Reply:
done